0.5 Changelog and documentation updates (#544)
PennyLaneAI/catalyst: doc/changelog.md

Reviewer: @josh146 The requirement to Rust compiler and toolchain in the installation page can be removed as PR https://github.com/PennyLaneAI/catalyst/pull/470 removes the dependency

Author: done in e8ac53b1

  Array(-0.93005586, dtype=float64)
  ```
  
  We can use `catalyst.vmap` to introduce additional batch dimensions to our input arguments,
  without needing to use a Python for loop:
  
  ```pycon
  >>> x = jnp.array([[0.1, 0.2, 0.3],
  ...                [0.4, 0.5, 0.6],
  ...                [0.7, 0.8, 0.9]])
  >>> y = jnp.array([jnp.pi, jnp.pi / 2, jnp.pi / 4])
  >>> qjit(vmap(cost))(x, y)
  array([-0.93005586, -0.97165424, -0.6987465 ])
  ```
  
  `catalyst.vmap()` has been implemented to match the same behaviour of `jax.vmap`, so should be a drop-in
  replacement in most cases. Under-the-hood, it is automatically inserting Catalyst-compatible for loops,
  which will be compiled and executed outside of Python for increased performance.

* Catalyst now supports compiling and executing QJIT-compiled QNodes using the
  CUDA Quantum compiler toolchain.
  [(#477)](https://github.com/PennyLaneAI/catalyst/pull/477)
  [(#536)](https://github.com/PennyLaneAI/catalyst/pull/536)
  [(#547)](https://github.com/PennyLaneAI/catalyst/pull/547)

  Simply import the CUDA Quantum `@cudaqjit` decorator to use this functionality:

  ```python
  from catalyst.cuda import cudaqjit
  ```

>   Or, if using Catalyst from PennyLane, simply specify `@qml.qjit(compiler="cuda_quantum")`.

Author: Yeah I had the same question 🤔 My current approach is:

- if I'm in the Catalyst docs, I use `@qjit`
- if I'm in the PL docs, I use `@qml.qjit`

But this is worth revisiting as we get better integration and feature parity

  The following devices are available when compiling with CUDA Quantum:

  * `softwareq.qpp`: a modern C++ statevector simulator
  * `nvidia.custatevec`: The NVIDIA CuStateVec GPU simulator (with support for multi-gpu)
  * `nvidia.cutensornet`: The NVIDIA CuTensorNet GPU simulator (with support for matrix product state)
  
  For example:

  ```python
  dev = qml.device("softwareq.qpp", wires=2)

  @cudaqjit
  @qml.qnode(dev)
  def circuit(x):
      qml.RX(x[0], wires=0)
      qml.RY(x[1], wires=1)
      qml.CNOT(wires=[0, 1])
      return qml.expval(qml.PauliY(0))
  ```

  ```pycon
  >>> circuit(jnp.array([0.5, 1.4]))
  -0.47244976756708373
  ```

  Note that CUDA Quantum compilation currently does not have feature parity with Catalyst

>   compilation; in particular, AutoGraph, control flow, differentiation, and various measurement

Author: No, limited classical code is supported. Let me add this.

>   statistics (such as probabilities and variance) are not yet supported.

Author: ```suggestion
  statistics (such as probabilities and variance) are not yet supported.
  Classical code support is also limited.
```

  ```

  ```pycon
  >>> f(1, MyClass(5))
  array(6)
  >>> f(1, MyClass(6))  # re-compilation
  array(7)
  >>> f(2, MyClass(5))  # no re-compilation
  array(7)
  ```

* Mid-circuit measurements now support post-selection and qubit reset when used with
  the Lightning simulators.
  [(#491)](https://github.com/PennyLaneAI/catalyst/pull/491)
  [(#507)](https://github.com/PennyLaneAI/catalyst/pull/507)

  To specify post-selection, simply pass the `postselect` argument to the `catalyst.measure`
  function:

  ```python
  dev = qml.device("lightning.qubit", wires=1)

  @qjit
  @qml.qnode(dev)
  def f():
      qml.Hadamard(0)
      m = measure(0, postselect=1)
      return qml.expval(qml.PauliZ(0))
  ```

  Likewise, to reset a wire after mid-circuit measurement, simply specify `reset=True`:

  ```python
  dev = qml.device("lightning.qubit", wires=1)

  @qjit
  @qml.qnode(dev)
  def f():
      qml.Hadamard(0)
      m = measure(0, reset=True)
      return qml.expval(qml.PauliZ(0))
  ```

<h3>Improvements</h3>

* Catalyst now supports Python 3.12
  [(#532)](https://github.com/PennyLaneAI/catalyst/pull/532)

* The JAX version used by Catalyst has been updated to `v0.4.23`.

>   [(#428)](https://github.com/PennyLaneAI/catalyst/pull/428)

Author: 🤔 I would say more an improvement than a breaking change (if we reserve the breaking change section for 'breaking changes to our API')

> * Catalyst now supports the differentiation of sliced arrays.

Author: Yeah, a lot of these entries could be both 😆 I moved a couple but left this one. Happy to move it down

> * Catalyst now supports the differentiation of sliced arrays.

Reviewer 2: Let's move it down then :)

  >>> catalyst.qjit(catalyst.grad(f))(x)
  [1. 0. 1. 0.]
  ```
* Catalyst now supports the `qml.GlobalPhase` operation.
  [(#563)](https://github.com/PennyLaneAI/catalyst/pull/563)

* Native support for `qml.PSWAP` and `qml.ISWAP` gates on Amazon Braket devices has been added.
  [(#458)](https://github.com/PennyLaneAI/catalyst/pull/458)

  Specifically, a circuit like

  ```py
  dev = qml.device("braket.local.qubit", wires=2, shots=100)

  @qjit
  @qml.qnode(dev)
  def f(x: float):
      qml.Hadamard(0)
      qml.PSWAP(x, wires=[0, 1])
      qml.ISWAP(wires=[1, 0])
      return qml.probs()
  ```

  would no longer decompose the `PSWAP` and `ISWAP` gates.

* The `qml.BlockEncode` operator is now supported with Catalyst.
  [(#483)](https://github.com/PennyLaneAI/catalyst/pull/483)

* Catalyst no longer relies on a TensorFlow installation for its AutoGraph functionality. Instead,
  the standalone `diastatic-malt` package is used and automatically installed as a dependency.
  [(#401)](https://github.com/PennyLaneAI/catalyst/pull/401)

> * The `@qjit` decorator will now remember previously compiled functions when the PyTree metadata
>   of arguments changes, in addition to already rememebering compiled functions when static
>   arguments change.

Reviewer 2: I might recommend something like this (also just noticed the spelling error)
```suggestion
* The `@qjit` decorator will remember previously compiled functions when the PyTree metadata
  of arguments changes, in addition to also remembering compiled functions when static
  arguments change.
```

  ```

  would no longer decompose the `PSWAP` and `ISWAP` gates.

* The `qml.BlockEncode` operator is now supported with Catalyst.
  [(#483)](https://github.com/PennyLaneAI/catalyst/pull/483)

* Additional debugging functions are now available in the `catalyst.debug` directory.
  [(#529)](https://github.com/PennyLaneAI/catalyst/pull/529)
  [(#522)](https://github.com/PennyLaneAI/catalyst/pull/531)

  This includes:

  - `compile_from_mlir(ir[, compiler, ...])` to compile a Catalyst function to binary code from the
    provided MLIR.

  - `filter_static_args(args, static_argnums)` to remove static values from arguments using the
    provided index list.

  - `get_cmain(fn, *args)` to return a C program that calls a jitted function with the provided
    arguments.

  - `print_compilation_stage(fn, stage)` to print one of the recorded compilation stages for a
    JIT-compiled function.

  - `promote_arguments(target_signature, args)` to promote arguments to the provided target
    signature, preserving PyTrees.

  For more details, please see the `catalyst.debug` documentation.

* Remove copies of TOML device configuration files for Lightning device in Catalyst.

>   [(#472)](https://github.com/PennyLaneAI/catalyst/pull/472)

Author: @erick-xanadu @maliasadi what does this entry mean?

Reviewer 3: There was a transition period where we needed copies of these files in the Catalyst repo. Now that the files live in PL-lightning, there's no reason to have them here.

Reviewer 3: ```suggestion
* Remove redundant copies of TOML files for `lightning.kokkos` and `lightning.qubit`.
  [(#472)](https://github.com/PennyLaneAI/catalyst/pull/472)
  
  `lightning.kokkos` and `lightning.qubit` now ship with their own TOML file. As such, we use the TOML file provided by them.
```

* Capturing quantum circuits with many gates prior to compilation is now quadratically faster (up to
  a factor), by removing `qextract_p` and `qinst_p` from forced-order primitives.
  [(#469)](https://github.com/PennyLaneAI/catalyst/pull/469)

* Update `AllocateQubit` and `AllocateQubits` in `LightningKokkosSimulator` to preserve
  the current state-vector before qubit re-allocations.

>   [(#479)](https://github.com/PennyLaneAI/catalyst/pull/479)

Reviewer: Ooops! I didn't get any GH notification for this comment 2d ago! :thinking: 

This PR is more like a technical update to the LightningKokkos support. It doesn't have any user impact as this is related to the dynamic qubits allocations/deallocations tickets. Secondary calls to `AllocateQubit` and `AllocateQubits` in LightningKokkosSimulator used to fully erase the previously allocated state-vector, but now second calls to these methods will update the previous state-vector to preserve the state at anytime within the scope of programs. This is not supported from the frontend and compiler at the moment and doesn't have any user impact. It can be also removed from the changelog, I guess.

> * Update `AllocateQubit` and `AllocateQubits` in `LightningKokkosSimulator` to preserve
>   the current state-vector before qubit re-allocations.
>   [(#479)](https://github.com/PennyLaneAI/catalyst/pull/479)

Reviewer: ```suggestion
* Update `AllocateQubit` and `AllocateQubits` in `LightningKokkosSimulator` to preserve
  the current state-vector before qubit re-allocations in the runtime dynamic qubits management.
  [(#479)](https://github.com/PennyLaneAI/catalyst/pull/479)
```

>   [(#479)](https://github.com/PennyLaneAI/catalyst/pull/479)

Reviewer 2: Using dynamic qubit allocation (not accessible to PennyLane users) used to erase the state of pre-existing qubits. This is no longer the case.

* The [PennyLane custom compiler entry point name convention has changed](https://github.com/PennyLaneAI/pennylane/pull/5140), necessitating
  a change to the Catalyst entry points.
  [(#493)](https://github.com/PennyLaneAI/catalyst/pull/493)

<h3>Breaking changes</h3>

* Catalyst gradient functions now match the Jax convention for the returned axes of
  gradients, Jacobians, VJPs, and JVPs. As a result, the returned tensor shape from various
  Catalyst gradient functions may differ compared to previous versions of Catalyst.
  [(#500)](https://github.com/PennyLaneAI/catalyst/pull/500)
  [(#501)](https://github.com/PennyLaneAI/catalyst/pull/501)
  [(#508)](https://github.com/PennyLaneAI/catalyst/pull/508)

> <h3>Internal changes</h3>

Author: @dime10 @maliasadi @erick-xanadu I added a new section to the changelog here, 'Internal Changes', because typically the Breaking Changes section refers to things that break in the 'public API' (e.g., user-facing Python functions, and device plugin API).

Whereas a lot of these entries seem more implementation details/internal changes with no implications on the public API

Reviewer 2: Seems like a good idea, but I think one those things could still be considered breaking the public API, which is moving the debug methods on the QJIT class.

Reviewer 2: But we also have other "public" interfaces to consider no?
- the QuantumDevice interface in the runtime for device plugins
- the QIR/C-API in the runtime for compilers
- the TOML schema for devices
- the Python entry points for frontends
- maybe others

I know they may not affect typical users, but they do affect integrators and aren't necessarily "internal" since they are meant for integration with other components or projects.

Author: Thanks @dime10 this is helpful :+1:

Author: I think it is up to us to make a decision here. I'm always a bit more conservative; we should err on something being internal _until_ we want to guarantee maintenance/not making breaking changes.

That is, the public API is what we declare to be the public API.

From the comments below, I'm inclined to say:

**Breaking change:**

- QuantumDevice interface (affects external plugin developers)
- TOML Schema

**Internal change:**
- Python entry points (since technically, the breaking change is in the PL API for compiler entrypoints, we are just conforming to it in this release)
- QIR/C-API in the runtime, based on Erick's comment below

Reviewer 3: I agree on this assessment :)

Reviewer: :100:

* The Catalyst Python frontend has been partially refactored. The impact on user-facing
  functionality is minimal, but the location of certain classes and methods used by the package
  may have changed.
  [(#529)](https://github.com/PennyLaneAI/catalyst/pull/529)
  [(#522)](https://github.com/PennyLaneAI/catalyst/pull/531)

  The following changes have been made:

  * Some debug methods and features on the QJIT class have been turned into free functions and moved
    to the `catalyst.debug` module, which will now appear in the public documention. This includes
    compiling a program from IR, obtaining a C program to invoke a compiled function from, and
    printing fine-grained MLIR compilation stages.

  * The `compilation_pipelines.py` module has been renamed to `jit.py`, and certain functionality
    has been moved out (see following items).

  * A new module `compiled_functions.py` now manages low-level access to compiled functions.

  * A new module `tracing/type_signatures.py` handles functionality related managing arguments
    and type signatures during the tracing process.

  * The `contexts.py` module has been moved from `utils` to the new `tracing` sub-module.

> * `QCtrl` is overriden and never used.

Reviewer 3: This has been reverted by Sergei. Effectively this change never happened.

  [{'expval0': (array(-0.0978434), array(-0.19767681))}, (array(-0.0978434), array(-0.19767681))]
  ```

> * Support has been added for linear algebra functions which depend on computing the eigenvalues
>   of symmetric matrices, such as `np.sqrt_matrix()`.

Reviewer: ```suggestion
* Support has been added for linear algebra functions which depend on computing the eigenvalues
  of symmetric matrices at run time backed by `lapack_dsyevd`, such as `np.sqrt_matrix()`.
```

Author: @maliasadi I moved the mention of lapack/DSYEVD below :)

>   of symmetric matrices, such as `np.sqrt_matrix()`.

Reviewer 2: Are there additional numpy functions that we know about in this category?

Author: @maliasadi?

Reviewer: @dime10 @josh146 I've tested the support with `np.eig` and `np.eigvals`. But there can be more linear algebra methods that depend on the calculation of eigenvalues under the hood.

  [(#488)](https://github.com/PennyLaneAI/catalyst/pull/488)

  For example, you can compile `qml.math.sqrt_matrix`:

  ```python
  @qml.qjit
  def workflow(A):
      B = qml.math.sqrt_matrix(A)
      return B @ A

>   ```

Author: @maliasadi I am a bit confused here, because A seems to be a NumPy array (and so `qml.math.sqrt_matrix()` would dispatch to `np.sqrt_matrix()`. That would imply it is run at tracing time, rather than being captured by JAX/JAXPR?

Reviewer: @josh146 This would lower to `custom_calls` only for matrix variables. If the matrix is known at compile time, `qml.math.sqrt_matrix(A)` can be calculated while tracing. however, for variable matrices (like this example, `A` is a run time parameter), JAX lowers the code to `custom_call` LAPACK functions.

Author: @maliasadi so this requires that `A` is a JAX array/runtime parameter?

Author: got it

Reviewer: yes!

  Internally, this involves support for lowering the eigenvectors/values computation lapack method
  `lapack_dsyevd` via `stablehlo.custom_call`.

> * Additional debugging functions are now available in the `catalyst.debug` directory.

Reviewer 2: I still think this could be considered a breaking change, these functions are not new, rather they have been moved from the `QJIT` class, which is part of the public API. (although they are better documented now)

Author: Okay, let me add an item here

Author: done

  [(#529)](https://github.com/PennyLaneAI/catalyst/pull/529)
  [(#522)](https://github.com/PennyLaneAI/catalyst/pull/531)

  This includes:

  - `compile_from_mlir(ir[, compiler, ...])` to compile a Catalyst function to binary code from the
    provided MLIR.

  - `filter_static_args(args, static_argnums)` to remove static values from arguments using the

>     provided index list.

Reviewer 2: This ...

Author: Ah, more utility functions? I'll rephrase

Reviewer 2: As these they need to be removed from this changelog entry which is about the debug functions (which are public).

These two are internal utility functions, they might not even deserve explicit mention beyond was is already mentioned in the refactor entry?

Author: done!

  - `get_cmain(fn, *args)` to return a C program that calls a jitted function with the provided
    arguments.

  - `print_compilation_stage(fn, stage)` to print one of the recorded compilation stages for a
    JIT-compiled function.

  - `promote_arguments(target_signature, args)` to promote arguments to the provided target

>     signature, preserving PyTrees.

Reviewer 2: ... and this are not debug functions. (they have been added to the new tracing module, and internal change)

  This is not so much a breaking change, it was just never used in the past.
  This will be set back to its original behaviour once it is implemented in the frontend.

> * Changes to the runtime QIR API and dependencies, to avoid symbol conflicts

Reviewer 3: It depends on whether we have consumers of this IR. The intended use case of Catalyst is that device implementors implement the `QuantumDevice` C++ interface and they don't have to worry about QIR. I consider this an internal change at the moment.

But once we start feeding QIR as an input to other devices, this would be considered a breaking change.

  * The Catalyst runtime now has a different API from QIR instructions.
  
    The runtime has been modified such that QIR instructions are lowered to functions where
    the `__quantum__` part of the function name is replaced with `__catalyst__`. This prevents
    the possibility of symbol conflicts with other libraries that implement QIR as a library.

  * The Catalyst runtime no longer depends on QIR runner's stdlib.

    We no longer depend nor link against QIR runner's stdlib. By linking against QIR runner's stdlib,
    some definitions persisted that may be different than ones used by third party implementors. To
    prevent symbol conflicts QIR runner's stdlib was removed and is no longer linked against. As a
    result, the following functions are now defined and implemented in Catalyst's runtime:

    * `int64_t __catalyst__rt__array_get_size_1d(QirArray *)`
    * `int8_t *__catalyst__rt__array_get_element_ptr_1d(QirArray *, int64_t)`

    and the following functions were removed since the frontend does not generate them

    * `QirString *__catalyst__rt__qubit_to_string(QUBIT *)`
    * `QirString *__catalyst__rt__result_to_string(RESULT *)`

<h3>Bug fixes</h3>

> 

Author: @erick-xanadu I can't comment on it, but what does the changelog item directly below (#555) mean?

Reviewer 3: Do you mean this?

>  Applying `map_wires` to `QCtrl` lead to an incorrect circuit. The wires in the nested regions remained unchanged.

What this means is that the implementation of `map_wires` (which `QCtrl` inherited from `Operation`) was wrong. This is because `QCtrl` contains a nested tape and did not apply `map_wires` to any of its nested tapes. All HybridOps should implement this function. I have added this ticket: https://github.com/PennyLaneAI/catalyst/issues/556

Reviewer 3: Does my comment answer your question?

Author: Yep! Although I'd be curious if this was a bug a user could access, and if so we could include a short example

Reviewer 3: I think definitely. Let me come up with an example.

Reviewer 3: ```python
import pennylane as qml
from pennylane.tape import QuantumTape
from catalyst import qjit, cond
from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union
from pennylane.transforms.core import transform

dev = qml.device("lightning.qubit", wires=2)

@transform
def custom_transform(tape: QuantumTape) -> (Sequence[QuantumTape], Callable):
    new_ops = []
    for op in tape:
        new_op = op.map_wires({0: 1, 1: 0})
        new_ops.append(new_op)

    new_tape = QuantumTape(new_ops, tape.measurements)
    return [new_tape], lambda res: res[0]

@qjit
@custom_transform
@qml.qnode(dev)
def qnn(c):
    @cond(c)
    def foo():
        qml.RX(3.14, wires=[1])
    foo()
    return qml.state()

print(qnn(True))
```

  of `target="mlir"`.
  [(#522)](https://github.com/PennyLaneAI/catalyst/pull/531)

* Fixes a bug in the configuration of dynamic shaped arrays that would cause certain program to
  error with `TypeError: cannot unpack non-iterable ShapedArray object`.
  [(#526)](https://github.com/PennyLaneAI/catalyst/pull/526)

  This is fixed by replacing the code which updates the `JAX_DYNAMIC_SHAPES` option with a
  `transient_jax_config()`` context manager which temporarily sets the value of
  `JAX_DYNAMIC_SHAPES` to True and then restores the original configuration value following the
  yield. The context manager is used by `trace_to_mlir()``.

* Exceptions encountered in the runtime when using the `@qjit` option `async_qnodes=Tue`
  will now be properly propagated to the frontend.
  [(#447)](https://github.com/PennyLaneAI/catalyst/pull/447)
  [(#510)](https://github.com/PennyLaneAI/catalyst/pull/510)

> 

Author: @dime10 I can't comment on it below, but I'm a bit confused about the `qinst` primitve entry; it says this change is not user facing, so I'm confused what the underlying bug was?

Reviewer 2: The bug was if present if you use some non-user functions, it's not a user-facing change because that functionality was not affected by the bug.

    * `QirString *__catalyst__rt__result_to_string(RESULT *)`

* Fix an issue when no qubit number was specified for the `qinst` primitive. The primitive now
  correctly deduces the number of qubits when no gate parameters are present. This change is not
  user facing.
  [(#496)](https://github.com/PennyLaneAI/catalyst/pull/496)

<h3>Bug fixes</h3>

* Fixed a bug where quantum control applied to a subcircuit was not correctly mapping wires,
  and the wires in the nested region remained unchanged.
  [(#555)](https://github.com/PennyLaneAI/catalyst/pull/555)

* Catalyst will no longer print a warning that recompilation is triggered when a `@qjit` decorated
  function with no arguments is invoke without having been compiled first, for example via the use
  of `target="mlir"`.
  [(#522)](https://github.com/PennyLaneAI/catalyst/pull/531)

* Fixes a bug in the configuration of dynamic shaped arrays that would cause certain program to
  error with `TypeError: cannot unpack non-iterable ShapedArray object`.
  [(#526)](https://github.com/PennyLaneAI/catalyst/pull/526)

  This is fixed by replacing the code which updates the `JAX_DYNAMIC_SHAPES` option with a

>   `transient_jax_config()`` context manager which temporarily sets the value of
>   `JAX_DYNAMIC_SHAPES` to True and then restores the original configuration value following the
>   yield. The context manager is used by `trace_to_mlir()``.

Reviewer 2: ```suggestion
  `transient_jax_config()` context manager which temporarily sets the value of
  `JAX_DYNAMIC_SHAPES` to True and then restores the original configuration value following the
  yield. The context manager is used by `trace_to_jaxpr()` and `lower_jaxpr_to_mlir()`.
```

  qubit re-allocation.

>   [(#473)](https://github.com/PennyLaneAI/catalyst/pull/473)

Author: @maliasadi what does this changelog entry mean?

Reviewer: There was a bug in the implementation of `AllocateQubits` (batch qubit allocations) in which second calls to this method would create more than required runtime qubit IDs. Similar to PR #479, this fix doesn't have any direct user impact.

  nested branches. The `wires` property in `adjoint` and `ctrl` cannot be used in workflows with
  control flow operations.

* Add support for lowering the eigenvectors/values computation lapack method: `lapack_dsyevd`
  via `stablehlo.custom_call`. For Catalyst, it means that you now can QJIT compile eigen
  vector/values operations and any other `qml.math` methods that uses these operations.

>   [(#488)](https://github.com/PennyLaneAI/catalyst/pull/488)

Author: @maliasadi I am a bit confused here, because `A` seems to be a NumPy array (and so `qml.math.sqrt_matrix()` would dispatch to `np.sqrt_matrix()`. That would imply it is run at tracing time, rather than being captured by JAX/JAXPR?

  * setting async runtime tokens and values to be errors
  * deallocating live tokens and values

> * Fix the scatter operation lowering when `updatedWindowsDim` is empty.

Reviewer 2: Do we know how this could have manifested for the user?

Author: Good question, cc @rmoyard

Reviewer 4: Possible for some indexing, but the main case is not covered because we removed the ability to differentiate with Jax when indexing. See https://github.com/PennyLaneAI/catalyst/pull/411#discussion_r1462447226

  [(#475)](https://github.com/PennyLaneAI/catalyst/pull/475)

> * Fix the issue in `LightningKokkos::AllocateQubits` with allocating too many qubit IDs on
>   qubit re-allocation.
>   [(#473)](https://github.com/PennyLaneAI/catalyst/pull/473)

Reviewer: ```suggestion
* Fix the issue in `LightningKokkos::AllocateQubits` with allocating too many qubit IDs on
  qubit re-allocation in the runtime dynamic qubits management support.
  [(#473)](https://github.com/PennyLaneAI/catalyst/pull/473)
```

* Fixed an issue where wires was incorrectly set as `<Wires = [<WiresEnum.AnyWires: -1>]>`
  when using `catalyst.adjoint` and `catalyst.ctrl`, by adding a `wires` property to
  these operations.
  [(#480)](https://github.com/PennyLaneAI/catalyst/pull/480)

  Currently, `catalyst.adjoint.wires` supports static wires and workflows with nested branches,
  and `catalyst.ctrl.wires` provides support for workflows with static and variable wires as well as
  nested branches. The `wires` property in `adjoint` and `ctrl` cannot be used in workflows with

>   control flow operations.

Author: This implies that there might be certain usages of `catalyst.adjoint`/`catalyst.ctrl` that are unsupported; do we have any examples we can include?

Reviewer 2: I think this is about the fact that dynamic wires will not be listed in the `.wires` property of the object, rather than limiting the use of `catalyst.adjoint/ctrl` (unless that use involves the wires property).

Author: Yeah, I would perhaps lean towards omitting them if they are more confusing rather than providing clarity...
